### PR TITLE
Telemetry frontend: metrics to stdout, tracing in http handlers, gRPC interceptor

### LINKED ISSUE #181: Instrument Frontend Service with Custom Metrics + Tracing
Subtask of #132 

- [ ] Add custom metrics for Nick's dashboard (HTTP request count, HTTP errors, HTTP request latency), export to Cloud Monitoring.
- [ ] Add tracing instrumentation to gRPC and HTTP handlers, export to Cloud Trace.

### PR DESCRIPTION
**Pull request by gzhao408**
Partially resolves #181, subtask of #132 

Progress so far:
- [x] Update docker image to be compatible with OpenTelemetry libraries
- [x] Add basic metrics to HTTP handlers
- [x] Add custom metrics (HTTP request count, HTTP errors count, HTTP request latency)
(https://github.com/GoogleCloudPlatform/stackdriver-sandbox/pull/183#issuecomment-664690690).
- [x] Add OTel gRPC interceptors to gRPC connections and the HTTP server
- [x] Export HTTP server traces to Cloud Trace

TODO:
- [ ] Testing script for tracing.
- [ ] Export metrics to Cloud Monitoring (blocked by dependency update in exporter, opened a [PR](https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/pull/73))

**How to test Metrics Manually:** Deploy (e.g. run `make deploy-continuous PROJECT_ID=<projectid>`) and observe what's printed to stdout from frontend service server. It'll look something like this, and obviously respond to requests to hipster shop.

image

<details><summary>Sample Output in Pretty JSON</summary>
<p>

```js
{

   "time":"2020-07-28T21:29:30.458881296Z",
   "updates":[

      {

         "name":"http_request_count{instrumentation.name=hipstershop/frontend}",
         "sum":5
      },
      {

         "name":"http.server.request_content_length{instrumentation.name=go.opentelemetry.io/otel/instrumentation/othttp,http.flavor=1.1,http.host=10.60.5.7:8080,http.scheme=http,http.server_name=frontend}",
         "sum":0
      },
      {

         "name":"http.server.request_content_length{instrumentation.name=go.opentelemetry.io/otel/instrumentation/othttp,http.flavor=1.1,http.host=34.83.167.77,http.scheme=http,http.server_name=frontend}",
         "sum":0
      },
      {

         "name":"http.server.duration{instrumentation.name=go.opentelemetry.io/otel/instrumentation/othttp,http.flavor=1.1,http.host=34.83.167.77,http.request_content_length=32,http.scheme=http,http.server_name=frontend}",
         "min":49211,
         "max":49211,
         "sum":49211,
         "count":1,
         "quantiles":[

            {

               "q":0.5,
               "v":49211
            },
            {

               "q":0.9,
               "v":49211
            },
            {

               "q":0.99,
               "v":49211
            }
         ]
      },
      {

         "name":"http.server.response_content_length{instrumentation.name=go.opentelemetry.io/otel/instrumentation/othttp,http.flavor=1.1,http.host=34.83.167.77,http.request_content_length=32,http.scheme=http,http.server_name=frontend}",
         "sum":0
      },
      {

         "name":"http.server.duration{instrumentation.name=go.opentelemetry.io/otel/instrumentation/othttp,http.flavor=1.1,http.host=34.83.167.77,http.scheme=http,http.server_name=frontend}",
         "min":38423,
         "max":118995,
         "sum":157418,
         "count":2,
         "quantiles":[

            {

               "q":0.5,
               "v":118995
            },
            {

               "q":0.9,
               "v":118995
            },
            {

               "q":0.99,
               "v":118995
            }
         ]
      },
      {

         "name":"http_response_errors{instrumentation.name=hipstershop/frontend}",
         "sum":1
      },
      {

         "name":"http.server.response_content_length{instrumentation.name=go.opentelemetry.io/otel/instrumentation/othttp,http.flavor=1.1,http.host=10.60.5.7:8080,http.scheme=http,http.server_name=frontend}",
         "sum":4
      },
      {

         "name":"http.server.duration{instrumentation.name=go.opentelemetry.io/otel/instrumentation/othttp,http.flavor=1.1,http.host=10.60.5.7:8080,http.scheme=http,http.server_name=frontend}",
         "min":178,
         "max":203,
         "sum":381,
         "count":2,
         "quantiles":[

            {

               "q":0.5,
               "v":203
            },
            {

               "q":0.9,
               "v":203
            },
            {

               "q":0.99,
               "v":203
            }
         ]
      },
      {

         "name":"http_request_latency{instrumentation.name=hipstershop/frontend}",
         "min":0,
         "max":119,
         "sum":207,
         "count":5,
         "quantiles":[

            {

               "q":0.5,
               "v":39
            },
            {

               "q":0.9,
               "v":119
            },
            {

               "q":0.99,
               "v":119
            }
         ]
      },
      {

         "name":"http.server.response_content_length{instrumentation.name=go.opentelemetry.io/otel/instrumentation/othttp,http.flavor=1.1,http.host=34.83.167.77,http.scheme=http,http.server_name=frontend}",
         "sum":22279
      },
      {

         "name":"http.server.request_content_length{instrumentation.name=go.opentelemetry.io/otel/instrumentation/othttp,http.flavor=1.1,http.host=34.83.167.77,http.request_content_length=32,http.scheme=http,http.server_name=frontend}",
         "sum":32
      }
   ]
}
```
</p>
</details>

**How to test Tracing Manually**:  Deploy (e.g. run `make deploy-continuous PROJECT_ID=<projectid>`) and then go to GCP Console > Cloud Trace > Trace List and filter using `LABEL:g.co/agent:opentelemetry`. You should see a number of traces with URI `span.frontend-server` with various latencies. 
image

Choose one with a higher latency (i.e. higher on the y-axis) and you'll see something interesting like this:
image

### CODE REVIEW

File: src/frontend/handlers.go [hunk 1]
```diff
@@ -88,6 +92,7 @@ func (fe *frontendServer) homeHandler(w http.ResponseWriter, r *http.Request) {
 
 func (fe *frontendServer) productHandler(w http.ResponseWriter, r *http.Request) {
 	log := r.Context().Value(ctxKeyLog{}).(logrus.FieldLogger)
+	trace.SpanFromContext(r.Context()).SetAttributes(kv.String("name", "product"))
```
Comment (simonz130) [inline src/frontend/handlers.go:95]: Is this required?
Can we "hide" it by using interceptors for all http incoming requests?

File: src/frontend/main.go [hunk 2]
```diff
@@ -31,6 +31,18 @@ import (
 	"go.opencensus.io/plugin/ochttp/propagation/b3"
 	"go.opencensus.io/stats/view"
 	"go.opencensus.io/trace"
+
+	texporter "github.com/GoogleCloudPlatform/opentelemetry-operations-go/exporter/trace"
```
Comment (simonz130) [inline src/frontend/main.go:35]: Worth adding a comment to explain which package are you using (exporter directly vs. collector and then exporting it)

File: src/frontend/main.go [hunk 3]
```diff
@@ -157,6 +184,17 @@ func initStats(log logrus.FieldLogger, exporter *stackdriver.Exporter) {
 	}
 }
 
+func initMeter(log logrus.FieldLogger) *push.Controller {
```
Comment (simonz130) [inline src/frontend/main.go:187]: document the method
also perhaps worth renaming to initMetricsCollection or something along those lines so it's clear what it does.

File: src/frontend/main.go [hunk 4]
```diff
@@ -157,6 +184,17 @@ func initStats(log logrus.FieldLogger, exporter *stackdriver.Exporter) {
...
+		Quantiles:   []float64{0.5, 0.9, 0.99},
+		PrettyPrint: false,
+	})
+	if err != nil {
+		log.Panicf("failed to initialize metric stdout exporter %v", err)
+	}
+	return pusher
+}
+
 func initStackdriverTracing(log logrus.FieldLogger) {
```
Comment (simonz130) [inline src/frontend/main.go:198]: This method name is not correct anymore :)

File: src/frontend/main.go [hunk 5]
```diff
@@ -187,6 +225,23 @@ func initTracing(log logrus.FieldLogger) {
...
+	exporter, err := texporter.NewExporter(texporter.WithProjectID(projectID))
+	if err != nil {
+		log.Fatalf("failed to initialize exporter: %v", err)
+	}
+
+	// Create trace provider with the exporter.
+	tp, err := sdktrace.NewProvider(sdktrace.WithConfig(
+		sdktrace.Config{DefaultSampler: sdktrace.AlwaysSample()}),
+		sdktrace.WithSyncer(exporter),
+		sdktrace.WithResource(resource.New(standard.ServiceNameKey.String("FrontendService"))))
```
Comment (simonz130) [inline src/frontend/main.go:240]: Isn't the package picking up the resource automatically? It should but it's possible that it's not implemented yet.
If it's not implemented yet, worth changing this to be "GKE". The resource name is supposed to be a platform on which the code is running.

File: src/frontend/main.go [hunk 6]
```diff
@@ -225,7 +280,9 @@ func mustConnGRPC(ctx context.Context, conn **grpc.ClientConn, addr string) {
 	*conn, err = grpc.DialContext(ctx, addr,
 		grpc.WithInsecure(),
 		grpc.WithTimeout(time.Second*3),
-		grpc.WithStatsHandler(&ocgrpc.ClientHandler{}))
+		grpc.WithStatsHandler(&ocgrpc.ClientHandler{}),
+		grpc.WithUnaryInterceptor(grpctrace.UnaryClientInterceptor(global.TraceProvider().Tracer("frontend"))),
+		grpc.WithStreamInterceptor(grpctrace.StreamClientInterceptor(global.TraceProvider().Tracer("frontend"))))
```
Comment (simonz130) [inline src/frontend/main.go:285]: I see you use interceptor. There should be a way to detect the name of the grpc action that is being called, so you shouldn't need to call 
```
trace.SpanFromContext(r.Context()).SetAttributes(kv.String("name", "home"))
```
and similar in each method

File: src/frontend/middleware.go [hunk 7]
```diff
@@ -58,6 +58,8 @@ func (lh *logHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	requestID, _ := uuid.NewRandom()
 	ctx = context.WithValue(ctx, ctxKeyRequestID{}, requestID.String())
 
+	http_request_count.Add(ctx, 1)
```
Comment (simonz130) [inline src/frontend/middleware.go:61]: There should be a built-in metric for instrumenting grpc with opentelemetry that should capture this. If this doesn't exist, then it's fine to leave it like this for now. Perhaps worth adding a //todo comment to remove it in the future when such metric is implemented in the core OT library.

File: src/frontend/main.go (hunk 6 above)
Comment (gzhao408) [reply src/frontend/main.go:286]: removed and added it to `ServeHTTP` with `r.URL.Path` instead

File: src/frontend/middleware.go (hunk 7 above)
Comment (gzhao408) [reply src/frontend/middleware.go:66]: There isn't a Views API and this isn't collected automatically, so it's custom metrics for now. With `othttp` it has some automatic ones but not the exact ones we want.

File: src/frontend/main.go (hunk 5 above)
Comment (gzhao408) [reply src/frontend/main.go]: Not 100% sure, but I changed it to GKE

File: src/frontend/main.go (hunk 4 above)
Comment (gzhao408) [reply src/frontend/main.go]: changed to `initOpenCensus` :)

File: src/frontend/main.go (hunk 3 above)
Comment (gzhao408) [reply src/frontend/main.go]: done

File: src/frontend/handlers.go (hunk 1 above)
Comment (gzhao408) [reply src/frontend/handlers.go]: hidden

File: src/frontend/main.go (hunk 2 above)
Comment (gzhao408) [reply src/frontend/main.go:34]: done

File: src/frontend/main.go [hunk 8]
```diff
@@ -186,7 +224,24 @@ func initTracing(log logrus.FieldLogger) {
 	// In a production environment or high QPS setup please use
 	// trace.ProbabilitySampler set at the desired probability.
 	trace.ApplyConfig(trace.Config{DefaultSampler: trace.AlwaysSample()})
-	initStackdriverTracing(log)
+	initOpenCensus(log)
```
Comment (simonz130) [inline src/frontend/main.go:227]: are you making a recursive call here?

**Comment by gzhao408** [comment]
This is what I can see in the Cloud Trace console:

image

I think this is the most I can get right now since I haven't instrumented any of the other services with OTel

File: src/frontend/main.go [hunk 9]
```diff
@@ -186,7 +224,24 @@ func initTracing(log logrus.FieldLogger) {
...
+
+	// Initialize OpenTelemetry Tracer
```
Comment (simonz130) [inline src/frontend/main.go:229]: I don't think the comment is right. You are initializing the exporter to Cloud Tracing (Google product, former Stackdriver Tracing)

**Comment by gzhao408** [comment]
Example of metrics stdout (includes othttp and the custom metrics):
<details><summary>Output</summary>
<p>

```js
{
   "time":"2020-07-27T23:18:24.31340596Z",
   "updates":[
      {
         "name":"http.server.request_content_length{instrumentation.name=go.opentelemetry.io/otel/instrumentation/othttp,http.flavor=1.1,http.host=10.60.3.7:8080,http.scheme=http,http.server_name=frontend}",
         "sum":0
      },
      {
         "name":"http.server.duration{instrumentation.name=go.opentelemetry.io/otel/instrumentation/othttp,http.flavor=1.1,http.host=10.60.3.7:8080,http.scheme=http,http.server_name=frontend}",
         "min":185,
         "max":273,
         "sum":458,
         "count":2,
         "quantiles":[
            {
               "q":0.5,
               "v":273
            },
            {
               "q":0.9,
               "v":273
            },
            {
               "q":0.99,
               "v":273
            }
         ]
      },
      {
         "name":"http.server.response_content_length{instrumentation.name=go.opentelemetry.io/otel/instrumentation/othttp,http.flavor=1.1,http.host=34.83.167.77,http.scheme=http,http.server_name=frontend}",
         "sum":30009
      },
      {
         "name":"http.server.response_content_length{instrumentation.name=go.opentelemetry.io/otel/instrumentation/othttp,http.flavor=1.1,http.host=34.83.167.77,http.request_content_length=32,http.scheme=http,http.server_name=frontend}",
         "sum":0
      },
      {
         "name":"http.server.response_content_length{instrumentation.name=go.opentelemetry.io/otel/instrumentation/othttp,http.flavor=1.1,http.host=10.60.3.7:8080,http.scheme=http,http.server_name=frontend}",
         "sum":4
      },
      {
         "name":"http.server.duration{instrumentation.name=go.opentelemetry.io/otel/instrumentation/othttp,http.flavor=1.1,http.host=34.83.167.77,http.scheme=http,http.server_name=frontend}",
         "min":7628,
         "max":26039,
         "sum":54705,
         "count":3,
         "quantiles":[
            {
               "q":0.5,
               "v":21038
            },
            {
               "q":0.9,
               "v":26039
            },
            {
               "q":0.99,
               "v":26039
            }
         ]
      },
      {
         "name":"http_request_latency{instrumentation.name=hipstershop/frontend}",
         "min":0,
         "max":26,
         "sum":59,
         "count":6,
         "quantiles":[
            {
               "q":0.5,
               "v":7
            },
            {
               "q":0.9,
               "v":26
            },
            {
               "q":0.99,
               "v":26
            }
         ]
      },
      {
         "name":"http.server.request_content_length{instrumentation.name=go.opentelemetry.io/otel/instrumentation/othttp,http.flavor=1.1,http.host=34.83.167.77,http.request_content_length=32,http.scheme=http,http.server_name=frontend}",
         "sum":32
      },
      {
         "name":"http.server.duration{instrumentation.name=go.opentelemetry.io/otel/instrumentation/othttp,http.flavor=1.1,http.host=34.83.167.77,http.request_content_length=32,http.scheme=http,http.server_name=frontend}",
         "min":4430,
         "max":4430,
         "sum":4430,
         "count":1,
         "quantiles":[
            {
               "q":0.5,
               "v":4430
            },
            {
               "q":0.9,
               "v":4430
            },
            {
               "q":0.99,
               "v":4430
            }
         ]
      },
      {
         "name":"http_request_count{instrumentation.name=hipstershop/frontend}",
         "sum":6
      },
      {
         "name":"http.server.request_content_length{instrumentation.name=go.opentelemetry.io/otel/instrumentation/othttp,http.flavor=1.1,http.host=34.83.167.77,http.scheme=http,http.server_name=frontend}",
         "sum":0
      },
      {
         "name":"http_response_errors{instrumentation.name=hipstershop/frontend}",
         "sum":2
      }
   ]
}
```
</p>
</details>

File: src/frontend/main.go [hunk 10]
```diff
@@ -186,7 +224,24 @@ func initTracing(log logrus.FieldLogger) {
...
+	exporter, err := texporter.NewExporter(texporter.WithProjectID(projectID))
+	if err != nil {
+		log.Fatalf("failed to initialize exporter: %v", err)
+	}
+
+	// Create trace provider with the exporter.
+	tp, err := sdktrace.NewProvider(sdktrace.WithConfig(
+		sdktrace.Config{DefaultSampler: sdktrace.AlwaysSample()}),
+		sdktrace.WithSyncer(exporter),
+		sdktrace.WithResource(resource.New(standard.ServiceNameKey.String("GKE"))))
```
Comment (simonz130) [inline src/frontend/main.go:240]: Do you know if they support resource auto-detection? Or at least a way to pass a pre-defined constant for GKE?
If both are not an option, at least please add a comment with TODO so when resource autodetection is supported, we should fix it here.

File: src/frontend/middleware.go [hunk 11]
```diff
@@ -74,6 +81,12 @@ func (lh *logHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 			"http.resp.took_ms": int64(time.Since(start) / time.Millisecond),
 			"http.resp.status":  rr.status,
 			"http.resp.bytes":   rr.b}).Debugf("request complete")
+		// TODO: remove when default metrics through Views API are avilable in OpenTelemetry
```
Comment (simonz130) [inline src/frontend/middleware.go:84]: nit: typo - available

File: src/frontend/main.go (hunk 8 above)
Comment (gzhao408) [reply src/frontend/main.go:227]: no - `main` calls `initTracing` which calls `initOpenCensus`

File: src/frontend/main.go (hunk 10 above)
Comment (gzhao408) [reply src/frontend/main.go:240]: added a TODO comment
there's no no predefined constant in the `standard` api
I'm 90% sure no auto-detection

File: src/frontend/main.go (hunk 9 above)
Comment (gzhao408) [reply src/frontend/main.go]: sorry yeah, updated

**Comment by gzhao408** [comment]
Rebased and squashed into 3 commits: docker image update, metrics + export to stdout, tracing + export to cloud trace
Also added some instruction on how to test manually in the PR description
Ready for review again!

File: src/frontend/main.go (hunk 10 above)
Comment (gzhao408) [reply src/frontend/main.go:245]: u were right I didn't need it :P

**Pull request merged**